adding kafka monitoring projects

assures opened the pull request:
Adding the code samples for librdkafka C# consumer and producer

arerlend commented on tutorials/monitoring/librdkafka/consumer/ConsumerLaglibrdkafka/appsettings.json:
```diff
@@ -0,0 +1,17 @@
+    }
+  },
+  "ApplicationInsights": {
+    "InstrumentationKey": "792b28cf-9f1a-4ec0-a040-824bf07c7a97"
```
secret?

assures replied:
InstrumentationKey left it with template string "XXXXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXXX. Let me know if it is fine

arerlend commented on tutorials/monitoring/librdkafka/producer/ProducerMetricsLibrdkafka/App.config:
```diff
@@ -0,0 +1,11 @@
+<configuration>
+  <appSettings>
+    <add key="EH_FQDN" value="evaskhub.servicebus.windows.net:9093"/>
+    <add key="EH_CONNECTION_STRING" value="Endpoint=sb://evaskhub.servicebus.windows.net/;SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=FLCFxOyQMmglHBvxG8XIkU7j4uuPZ4dQMRrpFRW3XvU="/>
```
secret

assures replied:
followed the quickstart sample to change values

arerlend commented on tutorials/monitoring/librdkafka/consumer/ConsumerLaglibrdkafka/Program.cs:
```diff
@@ -0,0 +1,124 @@
+                foreach (var topic in statistics.Topics)
+                {
+                    //getting the consumer lag value from all the paritions
+                    //You might see a parition id -1. This is for for internal UA/UnAssigned partition according to the doc here: https://github.com/edenhill/librdkafka/blob/master/STATISTICS.md
```
couple nits
- spelling partition
- unassigned (UA)
- leave spaces between // and the start of the comment

assures replied:
corrected the above three

arerlend commented on tutorials/monitoring/librdkafka/consumer/ConsumerLaglibrdkafka/Program.cs:
```diff
@@ -0,0 +1,124 @@
+                GroupId = consumerGroup,
+                AutoOffsetReset = AutoOffsetReset.Earliest,
+                //librdkafka statistics emit interval. Required to be configured for capturing the librdkafka statistics
+                StatisticsIntervalMs = 1000
```
Probably point this out in the readme, emitting telemetry every second can get very expensive when using App Insights

assures replied:
ok, will do

arerlend commented on tutorials/monitoring/librdkafka/consumer/ConsumerLaglibrdkafka/ConsumerLaglibrdkafka.csproj:
```diff
@@ -0,0 +1,25 @@
+
+  <ItemGroup>
+    <PackageReference Include="Confluent.Kafka" Version="1.5.1" />
+    <PackageReference Include="Microsoft.ApplicationInsights.WorkerService" Version="2.8.0-beta3" />
```
beta version? or should we reference stable version?  also for Microsoft.Extensions.Hosting, check that the version is a public build version, not from CoreXT on dev machine

assures replied:
changed the version from beta to 2.15.0. For Microsoft.Extensions.Hosting, version updated to stable 3.1.9. As a consequence, also upgraded System.Configuration.ConfigurationManage from 4.5.0 to 4.7.0 in order to work.

arerlend commented on tutorials/monitoring/librdkafka/consumer/ConsumerLaglibrdkafka/Program.cs:
```diff
@@ -0,0 +1,124 @@
+                    //Add the consumers as worker
+                    services.AddHostedService<Worker>();
+
+                    // Application Insights configs
```
floating comment

assures replied:
fixed

arerlend reviewed (commented):
Nice!  This will be very useful to customers.

Let's rename the project to DotnetConsumerLagMonitoringExample since there is really only one C# Kafka client.

Are we adding info in the README on how to implement this in other rdkafka clients?

assures commented on tutorials/monitoring/librdkafka/consumer/ConsumerLaglibrdkafka/App.config:
```diff
@@ -0,0 +1,11 @@
+<configuration>
+  <appSettings>
+    <add key="EH_FQDN" value="evaskhub.servicebus.windows.net:9093"/>
+    <add key="EH_CONNECTION_STRING" value="Endpoint=sb://evaskhub.servicebus.windows.net/;SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=FLCFxOyQMmglHBvxG8XIkU7j4uuPZ4dQMRrpFRW3XvU="/>
```
changed the keys/values as per the quickstart template

assures commented:
@arerlend, thanks, I have fixed the above points now. The directory structure also now changed to */quickstart/monitoring/librdkafka/dotnet/(consumer/producer)*.  Will include writeup for other librdkafka based clients in the README. Also maybe create a sample for that later if required.

arerlend commented:
Looks good, let's still add the section on handling -1 partition IDs in the code so that App Insights isn't overrun by -1 events.

Let's revisit once the README is added

assures commented:
@arerlend, just added README - under /dotnet and a root one under /monitoring. Will add writeup for other librdkafka clients in root README after a bit (still trying it out). Let me know if any comments for this, thx.

arerlend commented on tutorials/monitoring/java/README.md:
```diff
@@ -0,0 +1,118 @@
+
+For EH clients that are Java based, this document outlines an approach to monitor some client metrics, with the help of JMX. The monitoring will be done using Azure Monitor’s Application Insights. For this, we are recommending making use of the Application Insights Java 3.0 Preview. This is an agent which auto collects requests, dependencies and logs from the Java apps running anywhere on VMs, AK ...
+
+Start by downloading the agent from [here](https://github.com/microsoft/ApplicationInsights-Java/releases/download/3.0.0-PREVIEW.7/applicationinsights-agent-3.0.0-PREVIEW.7.jar). Further details regarding how to use the agent can be found [here](https://docs.microsoft.com/en-us/azure/azure-monitor/app/java-in-process-agent). 
```
Maybe we should link to this instead https://github.com/Microsoft/ApplicationInsights-Java/releases/latest

arerlend commented on tutorials/monitoring/java/README.md:
```diff
@@ -0,0 +1,118 @@
+
+The agent now will be able to collect the two JMX attributes shown above, namely the records-lag-max, which is the consumer lag and the records-consumed-rate.
+
+You can add additional metrics to the jmxMetrics list. There is also an additional field called heartbeat which is the frequency at which the metrics will be sent to the Application Insights Resource. A good choice would be 10 seconds. 
```
"Choosing a low heartbeat interval provides high fidelity monitoring but can result in a large AI bill..." or something like that

assures replied:
added the statement

arerlend commented on tutorials/monitoring/java/README.md:
```diff
@@ -0,0 +1,118 @@
+SetLocal
+set "KAFKA_JMX_OPTS=-Dcom.sun.management.jmxremote=true -Dcom.sun.management.jmxremote.authenticate=false -Dcom.sun.management.jmxremote.ssl=false -Djava.rmi.server.hostname=127.0.1.1 -Dcom.sun.management.jmxremote.rmi.port=9010"
+set "KAFKA_HEAP_OPTS=-Xmx512M"
+set "KAFKA_JVM_PERFORMANCE_OPTS=-javaagent:applicationinsights-agent-3.0.0-PREVIEW.5.jar"
```
Version doesn't match, should probably add a note for that

assures replied:
changed the version to matching one

arerlend commented on tutorials/monitoring/librdkafka/dotnet/producer/ProducerMetricsLibrdkafka/Program.cs:
```diff
@@ -0,0 +1,122 @@
+                {
+                    try
+                    {
+                        var dr = await producer.ProduceAsync("evaskhub", new Message<long, string> { Value = "test" });
```
EH name

assures replied:
fixed

arerlend commented on tutorials/monitoring/librdkafka/dotnet/producer/ProducerMetricsLibrdkafka/appsettings.Development.json:
```diff
@@ -0,0 +1,9 @@
+{
+  "Logging": {
+    "LogLevel": {
+      "Default": "Debug",
```
May be helpful to add a link to doc with log levels and description

arerlend commented on tutorials/monitoring/librdkafka/dotnet/producer/ProducerMetricsLibrdkafka/appsettings.json:
```diff
@@ -0,0 +1,17 @@
+      }
+    },
+    "LogLevel": {
+      "Default": "Information",
```
Which log level is actually used here?
